rfc/003-remote-pipelines/README.md

```diff
@@ -0,0 +1,206 @@
+# Remote Pipelines
+
+- **Status**: WIP
+- **Created**: Sep 14, 2022
+- **ETA**: Oct 14, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Discussion**: [PR #2577](https://github.com/tenzir/vast/pull/2577)
+
+## Overview
+
+This proposal enhances the local pipeline execution with a remote component,
+making it possible to connect pipeline across process boundaries.
+
+## Problem Statement
+
+As of RFC-001, VAST pipelines can execute locally. But a one-shot pipeline
+execution is not sufficient. After this building block, natural questions arise:
+
+- How can I connect a local and remote pipeline?
+- How can I manage (add/update/remove) a set of continuously running pipelines?
+- How do pipelines interact with VAST server nodes? Specifically, how can I send
+  data into a VAST node, and how can I query data from a VAST node?
+
+This proposal seeks an answer to these questions.
+
+## Solution Proposal
+
+The core of this proposal introduces a control plane that provides a global
+namespace for pipelines and VAST nodes. The ability to interconnect pipelines
+via unique names in location-independent manner enabeles a loosely coupled mesh
+arcitecture.
+
+The core primitives for establishing connections between pipelines are two
+dedicated operators, `push` and `pull`:
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `push`   | `Arrow`    | `Void`      | Sends data to a named pipeline
+| `pull`   | `Void`     | `Arrow`     | Attaches to a named pipeline
+
+Arguments to these operators represent names from the global namespace.
+
+### Example
+
+Assuming an existing mechanism to bind a pipeline to name, the new operators
+allow for interconnecting pipelines as follows.
+
+Create a named pipeline `pipe1` at a VAST node and expose in the global control
+plane:
+
+```bash
+vast spawn pipe1 'from - | read zeek | where orig_bytes > 1 GiB'
+vast expose pipe1 tcp://1.2.3.4:5555 
+```
+
+Attach to `pipe1` to read from it locally:
+
+```bash
+vast exec 'pull pipe1 | where resp_bytes > 1 KiB'
```

Comment:
This treats named pipelines as a second class construct.
How about getting rid of the `pull` and making name assignment part of the language instead of adding an extra command:

```
vast exec '
  let
    # local assignment 
    pipe1 = from - | read zeek | where orig_bytes > 1 GiB
  in 
    # exposing to the global namespace
    publish pipe1'

vast exec 'pipe1 | where resp_bytes > 1 KiB | write to lakeX as potential_exfiltration'

vast exec 'from /tmp/conn.log | pipe1'
```

PS: the `read zeek` should not be a thing, let's get rid of that.

Reply:
Interesting point. From a parser perspective `vast exec 'pull pipe1 | ...` seems easier to maintain than `vast exec 'pipe1 | ...`, because in the second scenario, your pipeline/dataflow (see comment above) names can conflict with operator names.